sui/manage-coins.js

```diff
@@ -0,0 +1,218 @@
+            }
+
+            if (coinsAtCursor.hasNextPage) {
+                cursor = coinsAtCursor.nextCursor;
+            }
+        } while (cursor);
+
+        return coinTypeToCoins;
+    }
+
+    static async splitCoins(tx, coinTypeToCoins, options) {
+        CoinManager.checkSplitAmount(options);
+        const splitAmount = BigInt(options.split);
+
+        console.log('\n==== Splitting Coins ====');
+
+        // Set coin type to given coin type or the first coin type if there's only one if it's a SUI token.
+        const hasOnlyGasToken = Object.keys(coinTypeToCoins).length === 1 && coinTypeToCoins[SUI_COIN_ID];
+        const coinType = options.coinType || hasOnlyGasToken ? SUI_COIN_ID : undefined;
+
+        if (coinType) {
+            // Throw an error if the coin type is specified but no coins are found
+            CoinManager.checkCoinType(coinType, coinTypeToCoins);
+            const coins = coinTypeToCoins[coinType];
+            const [coin] = CoinManager.doSplitCoins(tx, coins, splitAmount);
+
+            if (options.transfer) {
+                CoinManager.doTransfer(tx, coin, options.transfer);
+            }
+        } else {
+            for (const coinType in coinTypeToCoins) {
+                const coins = coinTypeToCoins[coinType];
+                if (this.isGasToken(coins.data[0])) continue;
+                const [coin] = CoinManager.doSplitCoins(tx, coins, splitAmount);
+
+                if (options.transfer) {
+                    CoinManager.doTransfer(tx, coin, options.transfer);
+                }
+            }
+        }
+
+        if (options.transfer) {
+            console.log(`\nTransfer ${splitAmount} coins for every split coin to ${chalk.green(options.transfer)}`);
+        }
+
+        // The transaction will fail if the gas budget is not set for splitting coins transaction
+        tx.setGasBudget(1e8);
+    }
+
+    static doTransfer(tx, coin, recipient) {
+        tx.transferObjects([coin], recipient);
+    }
+
+    static doSplitCoins(tx, coins, splitAmount) {
+        const firstObjectId = this.isGasToken(coins.data[0]) ? tx.gas : coins.data[0].coinObjectId;
+        const response = tx.splitCoins(firstObjectId, [splitAmount]);
+        console.log(`Split coins of type '${chalk.green(coins.data[0].coinType)}' with amount ${splitAmount}`);
+        return response;
+    }
+
+    static async mergeCoin(tx, coinTypeToCoins, options) {
+        const coinType = options.coinType;
+        console.log('\n==== Merging Coins ====');
+
+        // Throw an error if the coin type is specified but no coins are found
+        CoinManager.checkCoinType(coinType, coinTypeToCoins);
+
+        if (coinType) {
+            const coins = coinTypeToCoins[coinType];
+            await CoinManager.doMergeCoin(tx, coins);
+        } else {
+            for (const coinType in coinTypeToCoins) {
+                const coins = coinTypeToCoins[coinType];
```

Comment:
throw error if coins is undefined. Don't think a check helper is needed for it

Reply:
i think `coinType` cannot be undefined here, as it is being iterated over the defined keys in `coinTypeToCoins`.